File: Sharebook/Server/src/Services/Identity/Sharebook.Identity.API/Helpers/CredentialsHashCodeGenerator.cs
﻿namespace Sharebook.Identity.API.Helpers
{
    using Sharebook.Identity.API.Auth.Authentication;

    public static class CredentialsHashCodeGenerator
    {
        public static string CreateHash(IReadOnlyTwitterCredentials credentials)
        {
            var hash = $"{credentials.ConsumerKey} - {credentials.ConsumerSecret} - {credentials.BearerToken}";
            return $"{hash} - {credentials.AccessToken}  - {credentials.AccessTokenSecret}";
        }

        public static string CreateHash(IReadOnlyConsumerCredentials credentials)
        {
            if (credentials is IReadOnlyTwitterCredentials twitterCredentials)
            {
                return CreateHash(twitterCredentials);
            }

            return $"{credentials.ConsumerKey} - {credentials.ConsumerSecret} - {credentials.BearerToken}";
        }
    }
}

File: Sharebook/Client/src/app/core/Core/Helpers/CredentialsHashCodeGenerator.ts
import {IReadOnlyConsumerCredentials} from "../Models/Authentication/ReadOnlyConsumerCredentials";
import {IReadOnlyTwitterCredentials} from "../Models/Authentication/ReadOnlyTwitterCredentials";

export abstract class CredentialsHashCodeGenerator {
  public static createHash(credentials: IReadOnlyTwitterCredentials | IReadOnlyConsumerCredentials): string {
    if (CredentialsHashCodeGenerator.isIReadOnlyTwitterCredentials(credentials)) {
      let hash = `${credentials.consumerKey} - ${credentials.consumerSecret} - ${credentials.bearerToken}`;
      return `${hash} - ${credentials.accessToken}  - ${credentials.accessTokenSecret}`;
    }

    return `${credentials.consumerKey} - ${credentials.consumerSecret} - ${credentials.bearerToken}`;
  }

  private static isIReadOnlyTwitterCredentials(credentials: IReadOnlyConsumerCredentials): credentials is IReadOnlyTwitterCredentials {
    return (credentials as IReadOnlyTwitterCredentials).accessToken !== undefined;
  }
}
